Makefile

```diff
@@ -82,6 +82,19 @@ test: ebpf
 bench: ebpf
 	$(call build, build -F benchmark --release, building benchmarks)
 
+coverage: RUSTFLAGS += -C instrument-coverage
+coverage: ebpf
```

Comment:
> This is nice. We should try to extend it to all the builds regardless of whether it is a debug, release, or test (which is probably where it makes the most sense).

Do you really want to add it to `release`? I'm afraid this would make normal user confused for the profraw files.

For `test`, as `cargo test` does not build `target/debug/retis`, the `llvm-cov` report will fail.

> The only idea that comes to mind ATM is defining something like e.g. C=1 (or COV=1) instead of a dedicated target.

Yes, we can do this.

Reply:
> > This is nice. We should try to extend it to all the builds regardless of whether it is a debug, release, or test (which is probably where it makes the most sense).
> 
> Do you really want to add it to `release`? I'm afraid this would make normal user confused for the profraw files.
> 

that's a good point. Maybe we can exclude release (it's not really needed), but I would not special case it too much.
Given we're discussing it, it would be nice to have this for tests (see below) and the debug build, for release binaries it's not really useful, but maybe not that bad (assuming there's a risk it leads to ugly special cases).

> For `test`, as `cargo test` does not build `target/debug/retis`, the `llvm-cov` report will fail.

we don't, but usually, a binary gets generated and executed:
```
make test V=1
[...]
jobs=; CARGO_BUILD_JOBS=${jobs:-1} cargo   test                                                                                                                                                                                             
   Compiling retis v1.4.0 (/home/pvalerio/workspace/open_source/retis-org/retis/retis)                                                                                                                                                      
    Finished `test` profile [unoptimized + debuginfo] target(s) in 2.49s                                                                                                                                                                    
     Running unittests src/main.rs (target/debug/deps/retis-eed9ba1384fccece)                                                                                                                                                               
                                                                                                                                                                                                                                            
running 85 tests                                                                                                                                                                                                                            
test core::bpf_sys::bpf_sys::tests::load_xdp_insns ... ignored                                                                                                                                                                              
test collect::collector::tests::register_collectors ... ok                                                                                                                                                                                  
test collect::collector::tests::start_collectors ... ok                                                                                                                                                                                     
[...]
```
The problem is that the name changes.
Maybe this can help (didn't dig into it so requires a deeper look) for the `test ` case:

https://doc.rust-lang.org/cargo/reference/config.html#targetcfgrunner

> 
> > The only idea that comes to mind ATM is defining something like e.g. C=1 (or COV=1) instead of a dedicated target.
> 
> Yes, we can do this.

Reply:
It's weird that the `cargo test` report doesn't have coverage data, although the `profraw` files contain data. I use the following method
```
RUSTFLAGS="-C instrument-coverage" cargo test
llvm-profdata merge -sparse default_*.profraw -o retis_coverage.profdata
llvm-cov show --format=html --output-dir=coverage --instr-profile=retis_coverage.profdata target/debug/deps/retis-74bd6f5b55
```
And the index.html in coverage folder has 0 coverage result...

Reply:
> It's weird that the `cargo test` report doesn't have coverage data, although the `profraw` files contain data. I use the following method
> 
> ```
> RUSTFLAGS="-C instrument-coverage" cargo test
> llvm-profdata merge -sparse default_*.profraw -o retis_coverage.profdata

did you also try feeding `merge` with `retis/default_*.profraw`?

> llvm-cov show --format=html --output-dir=coverage --instr-profile=retis_coverage.profdata target/debug/deps/retis-74bd6f5b55
> ```
> 
> And the index.html in coverage folder has 0 coverage result...

Reply:
> did you also try feeding `merge` with `retis/default_*.profraw` (the workspace changed a bit the layouts)?

Yes, during the test I didn't add the retis prefix.

```
RUSTFLAGS="-C instrument-coverage" cargo test
llvm-profdata merge -sparse default_*.profraw -o retis_coverage.profdata
```

Reply:
In general, my suggestion here is to rely on wrappers.
A nice one is [tarpaulin](https://crates.io/crates/cargo-tarpaulin) but I think it is limited to test target (which IMO is already a nice thing).
The alternative is the one specified in #408 llvm-cov (`LLVM_COV=/usr/bin/llvm-cov LLVM_PROFDATA=/usr/bin/llvm-profdata cargo llvm-cov --lcov --output-path ./coverage.info test`) which allows you to instrument the build as well (probably, in this case, we could think of creating a `coverage-run` make target).
In both cases, we require some more dependencies (e.g. `cargo install cargo-llvm-cov`), but it's fine as they are consumed only upstream (tagging @atenart and @amorenoz keep them in the loop).

Both easily allow to generate reports, but also produce output formats supported by external services we might want to integrate with our CI ([coveralls](https://coveralls.io/) or [codecov](https://codecov.io/)). The idea is to extend the `test_script ` in our CI and upload coverage data to those services.

Lcov format should be fine in all the cases (I admit my env consumes that easily :) but it can be any.

Reply:
> It's weird that the `cargo test` report doesn't have coverage data, although the `profraw` files contain data. I use the following method

> And the index.html in coverage folder has 0 coverage result...

There are two issues to get coverage for tests:
1. The merging step should also include `*/default_*.profraw` files.
2. The test binary is not `target/debug/retis`. For each test flavor (unit tests, doc tests, etc) a dedicated binary is generated. They all look like `target/debug/deps/retis-c69a59fc9dc11e23` (I'm not sure how to get their exact hash or if it's possible to define one when building).

With the above we see coverage values for tests.

Reply:
> In general, my suggestion here is to rely on wrappers. A nice one is [tarpaulin](https://crates.io/crates/cargo-tarpaulin) 

I guess that would solve the "finding the test binary" issue?

> In both cases, we require some more dependencies (e.g. `cargo install cargo-llvm-cov`), but it's fine as they are consumed only upstream (adding @atenart and @amorenoz in the loop).

Agreed as this is only used for some optional targets.

> Both easily allow to generate reports, but also produce output formats supported by external services we might want to integrate with our CI ([coveralls](https://coveralls.io/) or [codecov](https://codecov.io/)). The idea is to extend the `test_script ` in our CI and upload coverage data to one of those.

Agreed to the above: we should get a result that is easy to integrate in the CI.

Reply:
> > In general, my suggestion here is to rely on wrappers. A nice one is [tarpaulin](https://crates.io/crates/cargo-tarpaulin)
> 
> I guess that would solve the "finding the test binary" issue?
>

Theoretically, a solution exists https://github.com/retis-org/retis/pull/408#discussion_r1665912167 and is aligned with your script proposal (would require at least a script).
My personal preference is toward using wrappers as, other than solving the test problem and other common ones, they produce only a very limited amount of output files.
 
If we need to instrument builds with CC, `cargo-llvm-cov` should better suit our needs see "[Get coverage of external tests](https://github.com/taiki-e/cargo-llvm-cov)" as well.

> > In both cases, we require some more dependencies (e.g. `cargo install cargo-llvm-cov`), but it's fine as they are consumed only upstream (adding @atenart and @amorenoz in the loop).
> 
> Agreed as this is only used for some optional targets.
> 
> > Both easily allow to generate reports, but also produce output formats supported by external services we might want to integrate with our CI ([coveralls](https://coveralls.io/) or [codecov](https://codecov.io/)). The idea is to extend the `test_script ` in our CI and upload coverage data to one of those.
> 
> Agreed to the above: we should get a result that is easy to integrate in the CI.

Reply:
> If we need to instrument builds with CC, `cargo-llvm-cov` should better suit our needs see "[Get coverage of external tests](https://github.com/taiki-e/cargo-llvm-cov?tab=readme-ov-file#get-coverage-of-external-tests)" as well.

Thanks for all your comments. `cargo-llvm-cov` looks good to me.  I will try it this week.